File: Data Structures & Algorithms/squares-of-a-sorted-array/submission-3.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        left = 0
        right = len(nums) - 1
        result = [0] * len(nums)
        write = len(nums) - 1

        while left <= right:
            if abs(nums[left]) > abs(nums[right]):
                result[write] = nums[left] * nums[left]
                left += 1
            else:
                result[write] = nums[right] * nums[right]
                right -= 1
            
            write -= 1
        return result
        

File: Data Structures & Algorithms/frequency-of-the-most-frequent-element/submission-3.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number}
     */
    //increment elements
    //sort the element
    maxFrequency(nums, k) {
        nums.sort((a,b)=> a - b);

        let left = 0;
        let currentSum = 0
        let maxFreq = 0
 
           // 2. Expand the window using the 'right' pointer
        for(let right = 0; right<nums.length; right++){
            currentSum += nums[right]

           // 3. Shrink the window from the left if the operations needed exceed k
        // Operations needed = (Window Width * Target Value) - Current Sum
            while((right - left + 1) * nums[right] -  currentSum > k){
                currentSum -= nums[left];
                left++
            };
            // 4. Update the maximum window size (frequency) found so far
            maxFreq = Math.max(maxFreq, right - left + 1)
        }
        return maxFreq;
    }
}
//Time Complexity: O(nlogn)
//Space Complexity: O(1)
